src/core/lib/security/transport/client_auth_filter.c

```diff
@@ -61,6 +68,40 @@ typedef struct {
   grpc_auth_context *auth_context;
 } channel_data;
 
+static void decode_cancel_state(gpr_atm cancel_state, grpc_closure **func,
+                                grpc_error **error) {
+  if (cancel_state & 1) {
+    *error = (grpc_error *)(cancel_state & ~(gpr_atm)1);
```

Comment:
BTW, it's also worth noting that this ugly atomic code is only temporary.  I'll be eliminating it as part of the call combiner work, which was my impetus for this PR in the first place.  If you want an early look, see this file in #11566 (which is built on top of this PR).